frontend/src/metabase/visualizations/echarts/cartesian/model/axis.ts

```diff
@@ -484,6 +492,160 @@ export const getXAxisEChartsType = (
   }
 };
 
+const getTickWithinRangePredicate = (range: DateRange) => {
+  const [minDate, maxDate] = range;
+
+  const isWidthinDataRange = (tickDateRaw: string) => {
+    const tickDate = dayjs(tickDateRaw);
+
+    if (minDate.isSame(maxDate)) {
+      return tickDate.isSame(minDate, "day");
+    }
+
+    const isBeforeRange = minDate && tickDate.isBefore(minDate);
+    const isAfterRange = maxDate && tickDate.isAfter(maxDate);
+
+    return !isBeforeRange && !isAfterRange;
+  };
+
+  return isWidthinDataRange;
+};
+
+const findRangeBasedTicksInterval = (
+  range: DateRange,
+): TimeSeriesInterval | null => {
+  // For ticks we want to pick the largest interval that exist 3 times in the range.
+  // For example, if data has week granularity but the range is more than 3 months, we want to show monthly ticks.
+  const ticksIntervalThreshold = 3;
```

Comment:
can be a constant outside of the function like `const TICKS_INTERVAL_THRESHOLD = 3`